src/libpgmoneta/security.c

```diff
@@ -3050,6 +3051,130 @@ pgmoneta_generate_string_hmac_sha256_hash(char* key, int key_length, char* value
+      0x63CD4B8FU, 0x91A6C88CU, 0x456CAC67U, 0xB7072F64U, 0xA457DC90U,
+      0x563C5F93U, 0x082F63B7U, 0xFA44E0B4U, 0xE9141340U, 0x1B7F9043U,
+      0xCFB5F4A8U, 0x3DDE77ABU, 0x2E8E845FU, 0xDCE5075CU, 0x92A8FC17U,
+      0x60C37F14U, 0x73938CE0U, 0x81F80FE3U, 0x55326B08U, 0xA759E80BU,
+      0xB4091BFFU, 0x466298FCU, 0x1871A4D8U, 0xEA1A27DBU, 0xF94AD42FU,
+      0x0B21572CU, 0xDFEB33C7U, 0x2D80B0C4U, 0x3ED04330U, 0xCCBBC033U,
+      0xA24BB5A6U, 0x502036A5U, 0x4370C551U, 0xB11B4652U, 0x65D122B9U,
+      0x97BAA1BAU, 0x84EA524EU, 0x7681D14DU, 0x2892ED69U, 0xDAF96E6AU,
+      0xC9A99D9EU, 0x3BC21E9DU, 0xEF087A76U, 0x1D63F975U, 0x0E330A81U,
+      0xFC588982U, 0xB21572C9U, 0x407EF1CAU, 0x532E023EU, 0xA145813DU,
+      0x758FE5D6U, 0x87E466D5U, 0x94B49521U, 0x66DF1622U, 0x38CC2A06U,
+      0xCAA7A905U, 0xD9F75AF1U, 0x2B9CD9F2U, 0xFF56BD19U, 0x0D3D3E1AU,
+      0x1E6DCDEEU, 0xEC064EEDU, 0xC38D26C4U, 0x31E6A5C7U, 0x22B65633U,
+      0xD0DDD530U, 0x0417B1DBU, 0xF67C32D8U, 0xE52CC12CU, 0x1747422FU,
+      0x49547E0BU, 0xBB3FFD08U, 0xA86F0EFCU, 0x5A048DFFU, 0x8ECEE914U,
+      0x7CA56A17U, 0x6FF599E3U, 0x9D9E1AE0U, 0xD3D3E1ABU, 0x21B862A8U,
+      0x32E8915CU, 0xC083125FU, 0x144976B4U, 0xE622F5B7U, 0xF5720643U,
+      0x07198540U, 0x590AB964U, 0xAB613A67U, 0xB831C993U, 0x4A5A4A90U,
+      0x9E902E7BU, 0x6CFBAD78U, 0x7FAB5E8CU, 0x8DC0DD8FU, 0xE330A81AU,
+      0x115B2B19U, 0x020BD8EDU, 0xF0605BEEU, 0x24AA3F05U, 0xD6C1BC06U,
+      0xC5914FF2U, 0x37FACCF1U, 0x69E9F0D5U, 0x9B8273D6U, 0x88D28022U,
+      0x7AB90321U, 0xAE7367CAU, 0x5C18E4C9U, 0x4F48173DU, 0xBD23943EU,
+      0xF36E6F75U, 0x0105EC76U, 0x12551F82U, 0xE03E9C81U, 0x34F4F86AU,
+      0xC69F7B69U, 0xD5CF889DU, 0x27A40B9EU, 0x79B737BAU, 0x8BDCB4B9U,
+      0x988C474DU, 0x6AE7C44EU, 0xBE2DA0A5U, 0x4C4623A6U, 0x5F16D052U,
+      0xAD7D5351U
+   };
+
+   for (unsigned int i = 0; i < size; i++)
+   {
+      crc32val = crc32_tab[((unsigned char) crc32val) ^
+                           ((unsigned char) buffer[i])] ^
+                 (crc32val >> 8);
+   }
+   *crc_buf = crc32val;
+   return 0;
+}
+
+int
+pgmoneta_create_crc32c_buffer_avx(char* buffer, size_t size, uint32_t* crc_buf)
+{
+   if (buffer == NULL)
+   {
+      return 1;
+   }
+   uint32_t crc32val = *crc_buf;
+
+   for (int i = 0; i < size; i++)
+   {
+      crc32val = _mm_crc32_u8(crc32val, *buffer++);
+   }
+   *crc_buf = crc32val;
+   return 0;
+}
+
+int
+pgmoneta_create_crc32c_file(char* path, char** crc)
+{
+   FILE* file = NULL;
+   char read_buf[16384];
+   unsigned long read_bytes = 0;
+   char* crc_string;
+   uint32_t crc_buf = 0;
+
+   file = fopen(path, "rb");
+   if (file == NULL)
+   {
+      return 1;
+   }
+
+   memset(read_buf, 0, sizeof(read_buf));
+
+   while ((read_bytes = fread(read_buf, 1, sizeof(read_buf), file)) > 0)
+   {
+      pgmoneta_create_crc32c_buffer(read_buf, sizeof(read_buf), &crc_buf);
+   }
+
+   fclose(file);
+
+   crc_string = malloc(9);
+   memset(crc_string, 0, 9);
```

Comment:
crc_string can be `NULL`

Reply:
Wont that lead to a segmentation fault when I use `sprintf`?

Reply:
```
if (crc_string == NULL)
{
   goto error;
}
```

and then

```
error:
   return 1;